### PR TITLE
Implement search resource using ElasticSearch

### PR DESCRIPTION
**Pull request by jirutka**
**TODO:**
- [x] write specs
- [x] fix searching room by room number without prefix
- [x] reindex after schedule planning (@flexik)

/cc @jnv review plz

### CODE REVIEW

**Comment by jnv** [comment]
Originally we have talked about using only an indexer working with a database. Any rationale for relying on ElasticSearch?

**Comment by jirutka** [comment]
I’ve eventually realized that it’s not possible (or it would be insane) to do this with PostgreSQL only.

**Comment by jnv** [comment]
I understand it but I am not very fond of having a dependency on such a humongous project for a tangential feature. Have you considered Sphinx or Xapian? Did you look at [Picky](https://github.com/floere/picky)? What's your integration testing strategy? Do you plan to mock-out ElasticSearch or do you want us to to actually have this monster running while doing tests?

**Comment by jirutka** [comment]
> for a tangential feature

This feature is extremely important for fittable.

> Have you considered Sphinx?

Yeah. I unfortunately haven’t found any usable library for Sphinx that doesn’t depend on ActiveRecord/Rails and didn’t have time to write my own. ES is an easier way. Also Sphinx isn’t as good as ES and it’s more useful for me to learn ES (that is based on ubiquitous Lucene) than Sphinx.

> Have you considered Xapian?

Nope, I haven’t found any mention about it when I’ve been doing research of search engines.

> Did you look at Picky?

Yes, I’ve been seriously considering it and discussed with @flexik and @deric. We all think that it’s not suitable for us. Maybe it would be, but it’s very badly documented, so can’t really say…

> Did you look at XYZ?

You can ask @deric about state of the art in lightweight search engines (for Ruby). There aren’t any options. :( [Ion](https://github.com/rstacruz/ion) seems to be most promising, but it lacks prefix search that is essential for fittable.

> What's your integration testing strategy?

Well, that’s the question I need to discuss with you.

> to actually have this monster running while doing tests?

We are already running full-blown PostgreSQL in all tests, because you haven’t properly isolated them, so… what’s the problem?

Look, I’ve tried very, very hard to get this work only with text-search support in PostgreSQL. It works (see branch [search-pg](https://github.com/cvut/sirius/commits/search-pg)) fine when searching only in IDs (codes), but it’s nearly impossible to mix it with multiword names (person and course) and get relevant ranking (read #91). PostgreSQL has higher than average text-search support, but it’s still “just” a SQL database.

File: spec/integration/elasticsearch/multi_search_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+  after(:context) do
+    MultiSearchIndex.delete
+  end
+
+
+  describe '.search' do
+
+    describe "finds course by" do
+
+      it "exact code" do
```
Comment (jnv) [inline spec/integration/elasticsearch/multi_search_index_spec.rb]: Use single quotation mark (unless you use string interpolation).

**Comment by jirutka** [comment]
I’m done (or at least I hope so). @jnv, please review it.

**Comment by jnv** [comment]
It looks fine, however… I think the development experience is not fine yet. I think we should have a Procfile for development which starts an elasticsearch cluster for you.

**Comment by jirutka** [comment]
> Procfile for development which starts an elasticsearch cluster for you.

Feel free to add it. ;)

File: spec/integration/elasticsearch/spec_helper.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+ELASTIC_ARGS = {
+  network_host: URI(Config.elastic_url).host,
+  port: URI(Config.elastic_url).port.to_s,
+  nodes: 1,
+  multicast_enabled: false
+}
+has_started_elastic = false
+
+def elastic_running?
```
Comment (jnv) [inline spec/integration/elasticsearch/spec_helper.rb]: This, along all those `before` contexts etc. should go to a separate support file.

**Comment by jnv** [comment]
There is no developer support in case of missing elasticsearch binary. Rake just dies with:

```
Failure/Error: Unable to find matching line from backtrace
     SystemExit:
       exit
```

**Comment by jnv** [comment]
Also, while you are at it, it should be specified what version of ElasticSearch I should use. Perhaps put a warning into `script/setup` and catch error if cluster fails to start (probably due to missing binary).

File: spec/integration/elasticsearch/spec_helper.rb (hunk 2 above)
Comment (jirutka) [inline spec/integration/elasticsearch/spec_helper.rb]: I’m confused… how do you imagine that?

**Comment by jirutka** [comment]
> Rake just dies with:
> ```
> Failure/Error: Unable to find matching line from backtrace
>     SystemExit:
>       exit
> ```

How do you run it? When I remove ES binaries and run it via rake, then I got this:

```
$ be rake spec:elasticsearch
I, [2015-07-06T17:07:59.971207 #24655]  INFO -- : ** [Raven] Raven 0.12.3 ready to catch errors
Run options: include {:focus=>true, :elasticsearch=>true}

MultiSearchIndex
Starting 1 Elasticsearch nodes..sh: elasticsearch: command not found
.........................^C
```

File: spec/integration/elasticsearch/spec_helper.rb (hunk 2 above)
Comment (jnv) [inline spec/integration/elasticsearch/spec_helper.rb]: Oh, nevermind, I didn't notice that you have this in a separate spec_helper already. This is fine.

**Comment by jnv** [comment]
I ran `be rake spec`. It is true I got:

```
MultiSearchIndex
Starting 1 Elasticsearch nodes..sh: 1: elasticsearch: not found
```

burried deep inside the tests; at the end I got only:

```
Failures:

  1) MultiSearchIndex.search courses by exact code: "MI-RUB" finds courses: "MI-RUB"
     Failure/Error: Unable to find matching line from backtrace
     SystemExit:
       exit
     # ./spec/integration/elasticsearch/spec_helper.rb:28:in `block (2 levels) in <top (required)>'
     # ./Rakefile:47:in `run_rspec!'
     # ./Rakefile:8:in `block (2 levels) in <top (required)>'
```

So this is specific to the RSpec output format. I wonder this can be improved, either skip those tests or change the captured failure.

**Comment by jnv** [comment]
Okay, the cluster start error is caused by Elasticsearch gem by calling `exit(1)` on its own. So just :shipit:, great work!

File: lib/paged.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+require 'ostruct'
+
+module Paged
```
Comment (jnv) [inline lib/paged.rb]: There is already `Paginate` interactor and `represent_paginated`. Why the hell there is a _third_ piece of code to do this – without explanation?

**Comment by jirutka** [comment]
Okay, I’m going to merge it… any final words, @jnv? :P

**Comment by jnv** [comment]
giphy

**Comment by kalvotom** [comment]
:-))))))

**Comment by jirutka** [comment]
@kalvotom Let’s [try](https://sirius.fit.cvut.cz/fittable-dev/) how awesome this search is! :smile_cat:

**Comment by kalvotom** [comment]
Cool!

**Comment by jirutka** [comment]
I’ve looked at the [Picky’s documentation](http://pickyrb.com/documentation.html) once again and now I have totally different feeling about it. :confused: Now I think that it would be well suitable for us. The only thing I can’t find in the documentation is how to customize score calculation when searching through multiple fields.

I think that the problem is that at the beginning I didn’t know what I really want, didn’t know theory and terminology behind text search. Therefore I’ve been quite confused from the Picky’s documentation. Then I learned some stuff when experimenting with text search in PostgreSQL and most of the stuff when learning ElasticSearch that has very good documentation and there are many blog posts about it.

Now I feel quite bad that I’ve condemned [this cute](http://pickyrb.com/images/picky-new.png) Picky and have chosen such enterprise beast as ElasticSearch. :crying_cat_face: Don’t get me wrong, ES works great and I like it (so far), but I still think that it’s quite a overkill for us. I’m usually trying to avoid too complex solutions, especially those written in Java and with “enterprise” label…

However, we will probably use ElasticSearch in other projects (e.g. logs collecting with Logstash or something similar), so it’s good to have experience with it.

**Pull request merged**